receiver/mysqlreceiver/scraper.go

```diff
@@ -0,0 +1,546 @@
+	dataPoint := metric.AppendEmpty()
+	dataPoint.SetTimestamp(ts)
+	dataPoint.SetIntVal(value)
+	if labels.Len() > 0 {
+		labels.CopyTo(dataPoint.Attributes())
+	}
+}
+
+// scrape scrapes the mysql db metric stats, transforms them and labels them into a metric slices.
+func (m *mySQLScraper) scrape(context.Context) (pdata.ResourceMetricsSlice, error) {
+
+	if m.client == nil {
+		return pdata.ResourceMetricsSlice{}, errors.New("failed to connect to http client")
+	}
+
+	// metric initialization
+	rms := pdata.NewResourceMetricsSlice()
+	ilm := rms.AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
+	ilm.InstrumentationLibrary().SetName("otel/mysql")
+	now := pdata.NewTimestampFromTime(time.Now())
+
+	bufferPoolPages := initMetric(ilm.Metrics(), metadata.M.MysqlBufferPoolPages).Gauge().DataPoints()
+	bufferPoolOperations := initMetric(ilm.Metrics(), metadata.M.MysqlBufferPoolOperations).Sum().DataPoints()
+	bufferPoolSize := initMetric(ilm.Metrics(), metadata.M.MysqlBufferPoolSize).Gauge().DataPoints()
+	commands := initMetric(ilm.Metrics(), metadata.M.MysqlCommands).Sum().DataPoints()
+	handlers := initMetric(ilm.Metrics(), metadata.M.MysqlHandlers).Sum().DataPoints()
+	doubleWrites := initMetric(ilm.Metrics(), metadata.M.MysqlDoubleWrites).Sum().DataPoints()
+	logOperations := initMetric(ilm.Metrics(), metadata.M.MysqlLogOperations).Sum().DataPoints()
+	operations := initMetric(ilm.Metrics(), metadata.M.MysqlOperations).Sum().DataPoints()
+	pageOperations := initMetric(ilm.Metrics(), metadata.M.MysqlPageOperations).Sum().DataPoints()
+	rowLocks := initMetric(ilm.Metrics(), metadata.M.MysqlRowLocks).Sum().DataPoints()
+	rowOperations := initMetric(ilm.Metrics(), metadata.M.MysqlRowOperations).Sum().DataPoints()
+	locks := initMetric(ilm.Metrics(), metadata.M.MysqlLocks).Sum().DataPoints()
+	sorts := initMetric(ilm.Metrics(), metadata.M.MysqlSorts).Sum().DataPoints()
+	threads := initMetric(ilm.Metrics(), metadata.M.MysqlThreads).Gauge().DataPoints()
+
+	// collect innodb metrics.
+	innodbStats, err := m.client.getInnodbStats()
+	for k, v := range innodbStats {
+		labels := pdata.NewAttributeMap()
+		switch k {
+		case "buffer_pool_size":
+			if f, ok := m.parseFloat(k, v); ok {
+				labels.Insert(metadata.L.BufferPoolSize, pdata.NewAttributeValueString("size"))
+				addToDoubleMetric(bufferPoolSize, labels, f, now)
+			}
+		}
+	}
+	if err != nil {
+		m.logger.Error("Failed to fetch InnoDB stats", zap.Error(err))
+		return pdata.ResourceMetricsSlice{}, err
+	}
+
+	// collect global status metrics.
+	globalStats, err := m.client.getGlobalStats()
+	if err != nil {
+		m.logger.Error("Failed to fetch global stats", zap.Error(err))
+		return pdata.ResourceMetricsSlice{}, err
+	}
+
+	for k, v := range globalStats {
+		labels := pdata.NewAttributeMap()
+		switch k {
+
+		// buffer_pool_pages
+		case "Innodb_buffer_pool_pages_data":
```

Comment:
The code below is very repetitive. Is it written this way for performance reasons? Otherwise, it seems like it could be factored into a part that extracts the magic strings from MySQL stats into a Go struct, and another that renders that struct into metrics.

The first part could be condensed a bit using the fact that each group of stats has a shared prefix. E.g., look for the prefix `Innodb_buffer_pool_pages_` and then use the rest of the string (`data`, `dirty`, `flushed` etc) as the attribute value.

Note that I haven't read all of the code in detail, so it's possible that there are special cases that break my simplistic proposal. Please feel free to ignore it, if that's the case :)

Reply:
It's very repetitive, and not for performance reasons. It's really just for simplicity and expediency.

In my experience, most metric scraping programs are conceptually very simple, but there is always a tendency to optimize for minimal lines of code. Abstractions are created that overfit for the current set of metrics, and then these abstractions become maintainability obstacles when an unanticipated exception reveals itself. I'm not saying this code is perfect, but I'd much rather it be simple and repetitive than unnecessarily clever.

I definitely think there is an opportunity to dry this code up, but I would like to approach it from a very broad perspective on scrapers in general. I have a sense that there's a helper package to be extracted that would standardize some or all of the standard scraping workflow of `raw result -> isolate value -> parse/process/nop -> map to output metric & labels` as well as standard handling of errors along the way.